forester-utils/src/rpc/solana_rpc.rs

```diff
@@ -272,116 +337,108 @@ impl RpcConnection for SolanaRpcConnection {
+                .map_err(RpcError::ClientError)?;
+            println!("Airdrop signature: {:?}", signature);
+            self.retry(|| async {
+                if self
+                    .client
+                    .confirm_transaction_with_commitment(&signature, self.client.commitment())?
+                    .value
+                {
+                    Ok(())
+                } else {
+                    Err(RpcError::CustomError("Airdrop not confirmed".into()))
                 }
-                Err(err) => {
-                    warn!("Airdrop request failed with error: {:?}", err);
-                }
-            }
-            tokio::time::sleep(Duration::from_secs(attempt as u64)).await;
-        }
-        Err(RpcError::CustomError(
-            "Max retries reached for airdrop request".into(),
-        ))
+            })
+            .await?;
+
+            Ok(signature)
+        })
+        .await
     }
 
     async fn get_balance(&mut self, pubkey: &Pubkey) -> Result<u64, RpcError> {
-        self.client.get_balance(pubkey).map_err(RpcError::from)
+        self.retry(|| async { self.client.get_balance(pubkey).map_err(RpcError::from) })
+            .await
     }
 
     async fn get_latest_blockhash(&mut self) -> Result<Hash, RpcError> {
-        self.client.get_latest_blockhash().map_err(RpcError::from)
+        self.retry(|| async { self.client.get_latest_blockhash().map_err(RpcError::from) })
+            .await
     }
 
     async fn get_slot(&mut self) -> Result<u64, RpcError> {
-        self.client.get_slot().map_err(RpcError::from)
+        self.retry(|| async { self.client.get_slot().map_err(RpcError::from) })
+            .await
     }
 
-    fn get_epoch_info(&self) -> Result<EpochInfo, RpcError> {
-        self.client.get_epoch_info().map_err(RpcError::from)
+    async fn warp_to_slot(&mut self, _slot: Slot) -> Result<(), RpcError> {
+        Err(RpcError::CustomError(
+            "Warp to slot is not supported in SolanaRpcConnection".to_string(),
+        ))
     }
 
     async fn send_transaction(&self, transaction: &Transaction) -> Result<Signature, RpcError> {
-        self.client
-            .send_transaction_with_config(
-                transaction,
-                RpcSendTransactionConfig {
-                    skip_preflight: true,
-                    max_retries: Some(3),
-                    ..Default::default()
-                },
-            )
-            .map_err(RpcError::from)
-    }
-
-    fn get_url(&self) -> String {
-        self.client.url()
+        self.retry(|| async {
+            self.client
+                .send_transaction_with_config(
+                    transaction,
+                    RpcSendTransactionConfig {
+                        skip_preflight: true,
+                        max_retries: Some(self.retry_config.max_retries as usize),
+                        ..Default::default()
+                    },
+                )
+                .map_err(RpcError::from)
```

Comment:
we use this logic in `send_transaction` in forester which already has retry logic, is that retry logic now redundant?

Reply:
Yes. We need retry logic for each rpc method (otherwise we can easily throw out of from process_queue), so I decided to consolidate it in one place.